### PR TITLE
Migrate tests to use public schema repo.

### PR DESCRIPTION
**Pull request by devunwired**
- Add actions-on-google/smart-home-schema as a submodule
- Redirect test logic to use appropriate schema files
- Fix broken tests

### CODE REVIEW

File: src/testlib.ts [hunk 1]
```diff
@@ -1,29 +1,80 @@
...
+/**
+ * Retrieve the JSON schema for the given intent.
+ * @param intentName FQN of the intent
+ * @param schemaType One of 'request' or 'response'
+ */
+function getIntentSchema(intentName: string, schemaType: string): object | undefined {
+  try {
+    const shortIntent = intentName.split('.').pop()?.toLowerCase() || 'unknown';
+    const filename = `${shortIntent}.${schemaType}.schema.json`;
+    return require(path.join('../schema/intents', shortIntent, filename));
```
Comment (proppy) [inline src/testlib.ts:19]: might be safer to use `__dirname` rather than `../` here: https://nodejs.org/docs/latest/api/modules.html#modules_dirname

Comment (proppy) [reply src/testlib.ts:19]: and JSON.parse

File: src/testlib.ts [hunk 2]
```diff
@@ -65,22 +117,23 @@ export function validate(intentRequest: object, apiResponse: object, syncData?:
       const attributes = syncDevices[i]['attributes'] || {};
       for (let j = 0; j < traits.length; j++) {
         const trait = traits[j];
-        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
-          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
+        const schema = getTraitSchema(trait,'attributes');
```
Comment (proppy) [inline src/testlib.ts:120]: nit: space

File: src/testlib.ts [hunk 3]
```diff
@@ -91,27 +144,29 @@ export function validate(intentRequest: object, apiResponse: object, syncData?:
...
-          const trait = syncDevices[j]['traits'];
-          if (trait in QUERY_COMMAND_STATES_EXPECT) {
-            const validateQueryTraitStates = responseValidation(states, QUERY_COMMAND_STATES_EXPECT[trait]);
-            if (validateQueryTraitStates) {
-              queryErrors.push(...validateQueryTraitStates);
+        const device = syncDevices.find((entry) => entry['id'] === deviceIds);
+        const traits = device['traits'];
+          for (let j = 0; j < traits.length; j++) {
+            const trait = traits[j];
+            const schema = getTraitSchema(trait,'states');
```
Comment (proppy) [inline src/testlib.ts:151]: nit: space

File: src/testlib.ts [hunk 4]
```diff
@@ -1,29 +1,80 @@
...
 
-const TRAIT_ATTRIBUTES_EXPECT = {
-  'action.devices.traits.OnOff': ON_OFF_ATTRIBUTES_SCHEMA,
-};
+/**
+ * Helper function to search for matching trait directory
+ * @param commandName
+ * @param schemaType
+ */
+function findEnclosingTrait(commandName: string, schemaType: string) {
```
Comment (proppy) [inline src/testlib.ts:65]: nit: might be easier to create that mapping on startup.

File: src/testlib.ts [hunk 5]
```diff
@@ -1,29 +1,80 @@
...
+  try {
+    const shortCommand = commandName.split('.').pop()?.toLowerCase() || 'unknown';
+    const filename = `${shortCommand}.${schemaType}.schema.json`;
+    const basePath = path.join(__dirname, '../schema/traits');
+    return fs.readdirSync(basePath).find((entry) =>
+      fs.readdirSync(path.join(basePath, entry)).includes(filename)
```
Comment (proppy) [inline src/testlib.ts:71]: rather than reading the whole dir you could just check that the file exist.

Comment (devunwired) [reply src/testlib.ts]: Good catch. Done.

File: src/testlib.ts (hunk 3 above)
Comment (devunwired) [reply src/testlib.ts]: Done

File: src/testlib.ts (hunk 2 above)
Comment (devunwired) [reply src/testlib.ts]: Done

File: src/testlib.ts (hunk 1 above)
Comment (devunwired) [reply src/testlib.ts]: Path fixed.

`require()` wraps the file read/cache/parse into one operation, and it was also the method previously used to load the schema files (keeping the change small). WDYT?

File: src/testlib.ts [hunk 6]
```diff
@@ -65,22 +119,23 @@ export function validate(intentRequest: object, apiResponse: object, syncData?:
       const attributes = syncDevices[i]['attributes'] || {};
       for (let j = 0; j < traits.length; j++) {
         const trait = traits[j];
-        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
-          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
+        const schema = getTraitSchema(trait, 'attributes');
```
Comment (proppy) [inline src/testlib.ts:122]: nit: maybe use a typescript enum for the schema type.

File: src/testlib.ts [hunk 7]
```diff
@@ -1,29 +1,82 @@
...
-  'action.devices.traits.OnOff': ON_OFF_ATTRIBUTES_SCHEMA,
-};
+/**
+ * Helper function to search for matching trait directory
+ * @param commandName
+ * @param schemaType
+ */
+function findEnclosingTrait(commandName: string, schemaType: string) {
+  try {
+    const shortCommand = commandName.split('.').pop()?.toLowerCase() || 'unknown';
```
Comment (proppy) [inline src/testlib.ts:69]: nit: maybe pass the short command as an argument?

File: src/testlib.ts [hunk 8]
```diff
@@ -1,29 +1,82 @@
...
-const DISCONNECT_RESPONSE_SCHEMA = require('../intents/disconnect.response.schema.json');
-const ON_OFF_STATES_SCHEMA = require('../traits/onoff.states.schema.json');
-const ON_OFF_ATTRIBUTES_SCHEMA = require('../traits/onoff.attributes.schema.json');
-
+/**
+ * Retrieve the JSON schema for the given intent.
+ * @param intentName FQN of the intent
+ * @param schemaType One of 'request' or 'response'
+ */
+function getIntentSchema(intentName: string, schemaType: string): object | undefined {
```
Comment (proppy) [inline src/testlib.ts:15]: nit: what about merging those into a `requireSchema` function, w/ a check/indirection on the type to compute the path?

File: src/testlib.ts [hunk 9]
```diff
@@ -1,29 +1,82 @@
...
+  try {
+    const shortIntent = intentName.split('.').pop()?.toLowerCase() || 'unknown';
+    const filename = `${shortIntent}.${schemaType}.schema.json`;
+    const basePath = path.join(__dirname, '../schema/intents');
```
Comment (proppy) [inline src/testlib.ts:19]: nit: what about folding the two path.join together?

**Review by proppy — APPROVED** [review]
Just some nits.

File: src/testlib.ts (hunk 8 above)
Comment (devunwired) [reply src/testlib.ts]: Done

File: src/testlib.ts (hunk 7 above)
Comment (devunwired) [reply src/testlib.ts:96]: In this case, both functions are called externally so it's helpful to keep the name parsing logic out of the main flow.

File: src/testlib.ts (hunk 6 above)
Comment (devunwired) [reply src/testlib.ts]: Done

**Pull request merged**